### PR TITLE
#660 New criteria system (API and integration into the flow)

### PR DESCRIPTION
**Pull request by rosiereflo**
also solve #676

### CODE REVIEW

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+    }
+  }
+
+  if (atLeastOneEligibleLoadWasFound && params_->getType() == criteria::CriteriaParams::SUM) {
+    if (params_->hasPMax() && sum > params_->getPMax()) {
+      Trace::debug() << DYNLog(SourcePowerAboveMax, sum, params_->getPMax(), params_->getId()) << Trace::endline;
+      return false;
+    }
+    if (params_->hasPMin() && sum < params_->getPMin()) {
+      Trace::debug() << DYNLog(SourcePowerBelowMin, sum, params_->getPMin(), params_->getId()) << Trace::endline;
```
Comment (adriguir) [inline dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: I have difficulties imagining a plausible criterion with this form with the current possibilities. Do you have in mind some possible examples? What is your opinion on this?

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+                Criteria(params) {}
+
+GeneratorCriteria::~GeneratorCriteria() {}
+
+bool
+GeneratorCriteria::checkCriteria(bool finalStep) const {
+  if (!finalStep && params_->getScope() == criteria::CriteriaParams::FINAL)
+    return true;
+  double sum = 0.;
+  bool atLeastOneEligibleLoadWasFound = false;
```
Comment (adriguir) [inline dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: Change to atLeastOneEligibleGeneratorWasFound

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp [hunk 3]
```diff
@@ -0,0 +1,213 @@
...
+BusCriteria::checkCriteria(bool finalStep) const {
+  assert(params_->getType() != criteria::CriteriaParams::SUM);
+  if (!finalStep && params_->getScope() == criteria::CriteriaParams::FINAL)
+    return true;
+  for (std::vector<boost::shared_ptr<BusInterface> >::const_iterator it = buses_.begin(), itEnd = buses_.end();
+      it != itEnd; ++it) {
+    double v = (*it)->getStateVarV();
+    if (doubleIsZero(v)) continue;
+    double vNom = (*it)->getVNom();
+    if (params_->hasUMaxPu() && v > params_->getUMaxPu()*vNom) {
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: It was precised nowhere so I took as reference the old code

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+  return true;
+}
+
+void
+BusCriteria::addBus(const boost::shared_ptr<BusInterface>& bus) {
+  if (params_->hasUMinNom() &&
+      bus->getVNom() < params_->getUMinNom()) return;
+  if (params_->hasUMaxNom() &&
+      bus->getVNom() > params_->getUMaxNom()) return;
+  if (doubleEquals(bus->getV0(), defaultV0) || bus->getV0() <  defaultV0) return;
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: this is 1., as if I remember well calculatedBus might have a V0 of 1. whereas they are not activated

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp (hunk 1 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: Yes, I agree this is a corner case, but I don't see a good reason to disable this possibility for generality reasons

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (adriguir) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: In the old implementation, it was 0.2:
  if (U0_ && v > 0 && vNom > 180 && U0_.value() > 0.2 && v < uMinPu * vNom) {
In which file is it defined right now?

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp (hunk 1 above)
Comment (adriguir) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: I was also hesitating what to do with this. Thus it is fine.

File: documentation/configuringDynawo/DynawoInputFiles.tex [hunk 5]
```diff
@@ -144,18 +144,91 @@ \subsection{par file}
...
 
 For each parameter, the user should give the name, the type (DOUBLE, INT, BOOL or STRING) and the value of the parameter. \\
 
 Sometimes, the par file contains references to parameters present in the iidm file (see paragraph \ref{sec:Set a reference to a value stored in the iidm file}).
 
+\subsection{crt file}
+
+The crt file contains all the criteria files to be checked during the
+simulation.
+\\
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply documentation/configuringDynawo/DynawoInputFiles.tex]: Solved

File: documentation/configuringDynawo/DynawoInputFiles.tex [hunk 6]
```diff
@@ -144,18 +144,91 @@ \subsection{par file}
...
+There are three different kind of criteria, depending on which type of component
+is targeted.
+
+Each criteria is then associated to an id, a set of parameters and optionally a list of component.
+
+The scope of a criteria (specified in its parameters) can be one of 'DYNAMIC' if
+the criteria should be checked at every time step or 'FINAL' if the criteria
+should be respected only at the final step.
+
+An empty component list means that all the component corresponding to the specified type will be subject to this
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply documentation/configuringDynawo/DynawoInputFiles.tex]: Solved (I learned something today ;))

File: documentation/resources/syntaxExample/criteria.crt [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+    <component id="_BUS___10_TN"/>
+  </busCriteria>
+  <busCriteria>
+    <parameters id ="maximum bus voltage with UNom above 180kV " scope="DYNAMIC" type="LOCAL_VALUE" uMinNom="180" uMaxPu="1.2"/>
+  </busCriteria>
+  <loadCriteria>
+    <parameters id ="maximum sum of load active power" scope="DYNAMIC" type="SUM" pMax="100" uMaxPu="0.6"/>
+  </loadCriteria>
+  <generatorCriteria>
+    <parameters id ="maximum sum of generator active power" scope="DYNAMIC" type="SUM" pMax="100" uMaxPu="0.6" uMinNom="180"/>
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply documentation/resources/syntaxExample/criteria.crt]: Solved

File: dynawo/sources/API/CRT/CMakeLists.txt [hunk 8]
```diff
@@ -0,0 +1,63 @@
+# Copyright (c) 2015-2019, RTE (http://www.rte-france.com)
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CMakeLists.txt]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaCollection.h [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+class CriteriaCollection {
+ public:
+  /**
+  * define type of components
+  */
+  typedef enum {
+    BUS,
+    LOAD,
+    GENERATORS
+  } CriteriaCollectionType_t;  ///< components type
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaCollection.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaCollection.h [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+ protected:
+  class BaseConstCriteriaCollectionIteratorImpl;  // Abstract class for the interface
+
+ public:
+  /**
+   * @class CriteriaCollectionConstIterator
+   * @brief Const iterator over curves
+   *
+   * Const iterator over curves stored in collection
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaCollection.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaCollectionFactory.cpp [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, you can obtain one at http://mozilla.org/MPL/2.0/.
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file  CRVCurvesCollectionFactory.cpp
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaCollectionFactory.cpp]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaCollectionImpl.cpp [hunk 12]
```diff
@@ -0,0 +1,142 @@
...
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file  CRVCurvesCollectionImpl.cpp
+ *
+ * @brief Curves collection : implementation file
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaCollectionImpl.cpp]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaCollectionImpl.cpp [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+  case CriteriaCollection::BUS:
+    busCriterias_.push_back(criteria);
+    break;
+  case CriteriaCollection::LOAD:
+    loadCriterias_.push_back(criteria);
+    break;
+  case CriteriaCollection::GENERATORS:
+    generatorCriterias_.push_back(criteria);
+    break;
+  }
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaCollectionImpl.cpp]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaCollectionImpl.h [hunk 14]
```diff
@@ -0,0 +1,190 @@
...
+/**
+ * @class CriteriaCollection::BaseConstCriteriaCollectionIteratorImpl
+ * @brief Implementation class for iterators' functions
+ */
+class CriteriaCollection::BaseConstCriteriaCollectionIteratorImpl {
+ public:
+  /**
+   * @brief Constructor
+   *
+   * Constructor based on curvesCollection. Can create an implementation for
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaCollectionImpl.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParams.h [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+   * @brief return true if uMax has been defined
+   * @return true if uMax has been defined
+   */
+  virtual bool hasUMaxPu() const = 0;
+
+  /**
+   * @brief Setter for maximum nominal voltage
+   * @param uMaxNom maximum nominal voltage
+   */
+  virtual void setUMaxNom(double uMaxNom) = 0;
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParams.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParams.h [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+   * @brief return true if uMin has been defined
+   * @return true if uMin has been defined
+   */
+  virtual bool hasUMinPu() const = 0;
+
+  /**
+   * @brief Setter for minimum nominal voltage
+   * @param uMinNom minimum nominal voltage
+   */
+  virtual void setUMinNom(double uMinNom) = 0;
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParams.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParams.h [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+   * @brief return true if uMaxNom has been defined
+   * @return true if uMaxNom has been defined
+   */
+  virtual bool hasUMaxNom() const = 0;
+
+  /**
+   * @brief Setter for minimum voltage
+   * @param uMin minimum voltage
+   */
+  virtual void setUMinPu(double uMin) = 0;
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParams.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParamsImpl.cpp [hunk 18]
```diff
@@ -0,0 +1,151 @@
...
+  id_ = id;
+}
+
+const std::string&
+CriteriaParams::Impl::getId() const {
+  return id_;
+}
+
+void
+CriteriaParams::Impl::setUMaxPu(double uMax) {
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParamsImpl.cpp]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParamsImpl.cpp [hunk 19]
```diff
@@ -0,0 +1,151 @@
...
+  return uMaxNom_;
+}
+
+bool
+CriteriaParams::Impl::hasUMaxNom() const {
+  return uMaxNom_ < std::numeric_limits<double>::max();
+}
+
+void
+CriteriaParams::Impl::setUMinPu(double uMin) {
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParamsImpl.cpp]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParamsImpl.h [hunk 20]
```diff
@@ -0,0 +1,172 @@
...
+
+  /**
+   * @copydoc CriteriaParams::hasUMaxNom() const
+   */
+  bool hasUMaxNom() const;
+
+  /**
+   * @copydoc CriteriaParams::setUMinPu(double uMin)
+   */
+  void setUMinPu(double uMin);
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParamsImpl.h]: Solved

File: dynawo/sources/API/CRT/CRTCriteriaParamsImpl.h [hunk 21]
```diff
@@ -0,0 +1,172 @@
...
+   */
+  bool hasPMin() const;
+
+ private:
+  CriteriaParams::CriteriaScope_t scope_;  ///< scope of the criteria
+  CriteriaType_t type_;  ///< type of the criteria
+  double uMinPu_;  ///< minimum voltage
+  double uMaxPu_;  ///< maximum voltage
+  double uMinNom_;  ///< minimum nominal voltage
+  double uMaxNom_;  ///< maximum nominal voltage
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTCriteriaParamsImpl.h]: Solved

File: dynawo/sources/API/CRT/CRTXmlHandler.cpp [hunk 22]
```diff
@@ -0,0 +1,157 @@
...
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file  CRVXmlHandler.cpp
+ *
+ * @brief Handler for curves collection file : implementation file
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTXmlHandler.cpp]: Solved

File: dynawo/sources/API/CRT/CRTXmlImporter.cpp [hunk 23]
```diff
@@ -0,0 +1,72 @@
...
+  try {
+    return importFromStream(stream);
+  } catch (const DYN::Error& exp) {
+    throw DYNError(DYN::Error::API, XmlFileParsingError, fileName, exp.what());
+  }
+}
+
+boost::shared_ptr<CriteriaCollection>
+XmlImporter::importFromStream(std::istream& stream) const {
+  XmlHandler curvesHandler;
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTXmlImporter.cpp]: Solved

File: dynawo/sources/API/CRT/CRTXmlImporter.cpp [hunk 24]
```diff
@@ -0,0 +1,72 @@
...
+
+  xml::sax::parser::ParserFactory parser_factory;
+  xml::sax::parser::ParserPtr parser = parser_factory.createParser();
+
+  try {
+    bool xsdValidation = false;
+    if (getEnvVar("DYNAWO_USE_XSD_VALIDATION") == "true") {
+      const std::string crvXsdPath = getMandatoryEnvVar("DYNAWO_XSD_DIR") + std::string("criteria.xsd");
+      parser->addXmlSchema(crvXsdPath);
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTXmlImporter.cpp]: Solved

File: dynawo/sources/API/CRT/CRTXmlImporter.h [hunk 25]
```diff
@@ -0,0 +1,53 @@
...
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file  CRVXmlImporter.h
+ *
+ * @brief Curves collection XML importer : header file
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/CRTXmlImporter.h]: Solved

File: dynawo/sources/API/CRT/test/TestCriteria.cpp [hunk 26]
```diff
@@ -0,0 +1,187 @@
...
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file API/CRV/test/TestCurve.cpp
+ * @brief Unit tests for API_CRV
+ *
+ */
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/test/TestCriteria.cpp]: Solved

File: dynawo/sources/API/CRT/test/TestXmlImporter.cpp [hunk 27]
```diff
@@ -0,0 +1,147 @@
...
+// SPDX-License-Identifier: MPL-2.0
+//
+// This file is part of Dynawo, an hybrid C++/Modelica open source time domain
+// simulation tool for power systems.
+//
+
+/**
+ * @file API/CRV/test/TestXmlImporter.cpp
+ * @brief Unit tests for API_CRV
+ *
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/API/CRT/test/TestXmlImporter.cpp]: Solved

File: dynawo/sources/Common/Dictionaries/DYNLog_en_GB.dic [hunk 28]
```diff
@@ -99,9 +97,16 @@ TapChangerLocked              =             %1%:  Tap changer is blocked
 NetworkInitSwitchCurrentsFailed =           model network : initialization of switches' currents failed
 //---------------------- DATAINTERFACE --------------------------------------
 VariableNotSet                =             %1% : "%3%" is not set for model %2%
-BusUnderVoltage               =             node: %1% has a voltage under 0.8Un
-LoadUnderVoltageT             =             total load power with voltage under 0.85Un = %1% MW > 200 MW
-LoadUnderVoltageTEnd          =             total load power with voltage under 0.92Un at the end of simulation = %1%  MW > 1500 MW
+BusUnderVoltage               =             node: %1% has a voltage %2% < %3%Un (criteria id: %4%)
+BusAboveVoltage               =             node: %1% has a voltage %2% > %3%Un (criteria id: %4%)
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Common/Dictionaries/DYNLog_en_GB.dic]: Solved

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp (hunk 2 above)
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: Solved

File: documentation/configuringDynawo/DynawoInputFiles.tex [hunk 29]
```diff
@@ -144,18 +144,91 @@ \subsection{par file}
...
+simulation.
+\\
+
+Here is an example of a crt file :
+\lstinputlisting[language=XML, morekeywords={criteria, parameters, component,
+busCriteria, loadCriteria, generatorCriteria}, deletekeywords={type}, title=crt
+file example]{../resources/syntaxExample/criteria.crt}
+
+There are three different kind of criteria, depending on which type of component
+is targeted.
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply documentation/configuringDynawo/DynawoInputFiles.tex]: Solved

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: DYNCommonConstants.h in DataInterface

As shown below, if U0 is not defined on a calculated bus its default value is defaultV0:
double
CalculatedBusInterfaceIIDM::getV0() const {
  if (U0_)
    return U0_.value();
  else
    return defaultV0;  // default value
}

**Comment by rosiereflo** [comment]
Warning message when no IIDM is defined is not in the update, I'm working on it.

File: dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (rosiereflo) [reply dynawo/sources/Modeler/DataInterface/DYNCriteria.cpp]: (I removed the 0.2 as it seemed to be arbitrarily chosen)

**Comment by rosiereflo** [comment]
> Warning message when no IIDM is defined is not in the update, I'm working on it.

Done

File: nrt/data/IEEE14/IEEE14_BasicTestCases/IEEE14_DisconnectGroup/IEEE14.jobs [hunk 30]
```diff
@@ -1,4 +1,4 @@
-<?xml version='1.0' encoding='UTF-8'?>
+<?xml version="1.0"?>
```
[the remark this answers is not in the document]
Comment (rosiereflo) [reply nrt/data/IEEE14/IEEE14_BasicTestCases/IEEE14_DisconnectGroup/IEEE14.jobs]: the xsl did this and I forgot to fix it. It is now solved.

**Pull request merged**